### PR TITLE
Provide ammo accessors for vehicle parts

### PR DESCRIPTION
**Pull request by mugling**
Encapsulate `vehicle_part::amount` via `ammo_*` accessors in preparation for generic vehicle tanks.

The plan is to mimic the interface of `item` as most operations will eventually defer to the `base` item which itself will store the fuel/charges. Incidental advantages of this approach would include simplifying installation/removal of parts and calculation of vehicle weight.

Testing and code review welcomed.

### CODE REVIEW

**Comment by Coolthulhu** [comment]
For some reason my (green) battery is leaking.

File: src/vehicle.cpp [hunk 1]
```diff
@@ -4195,34 +4139,30 @@ void vehicle::alarm() {
...
-    for( auto &elem : fuel ) {
-        vehicle_part &part = parts[elem];
-        vpart_info pinfo = part_info( elem );
-        if( pinfo.fuel_type != fuel_type_gasoline && pinfo.fuel_type != fuel_type_diesel &&
-            pinfo.fuel_type != fuel_type_battery && pinfo.fuel_type != fuel_type_water ) {
-            // Not a liquid fuel or battery
+    // for each badly damaged tanks (lower than 50% health), leak a small amount
+    for( auto &p : parts ) {
+        auto dmg = double( p.hp ) / p.info().durability;
+        if( dmg < 0.5 || p.ammo_remaining() <= 0 ) {
```
Comment (Coolthulhu) [inline src/vehicle.cpp]: Inverted logic due to bad choice of name: `dmg` means hp ratio, `dmg < 0.5` means "leak only from parts that are at 50% or less hp".

File: src/vehicle.cpp [hunk 2]
```diff
@@ -5806,7 +5746,7 @@ bool vehicle::explode_fuel( int p, damage_type type )
             name.c_str());
         g->explosion( global_part_pos3( p ), pow, 0.7, data.fiery_explosion );
         parts[p].hp = 0;
-        parts[p].amount = 0;
+        parts[p].ammo_set( "null", 0 );
```
Comment (Coolthulhu) [inline src/vehicle.cpp]: Why `ammo_set( "null", 0 )` instead of cleaner `ammo_unset`?

File: src/vehicle.cpp [hunk 3]
```diff
@@ -6923,6 +6845,53 @@ std::string vehicle_part::name() const {
...
+
+int vehicle_part::ammo_set( const itype_id &ammo, long qty )
+{
+    // @todo currently only support fuel tanks and batteries
+    if( !info().has_flag( VPFLAG_FUEL_TANK ) ) {
+        return -1;
+    }
+
+    if( ammo == "null" && qty == 0 ) {
+        return amount = 0;
```
Comment (Coolthulhu) [inline src/vehicle.cpp]: I don't like constructs like that. Saves one line, but adds one "WTF".
Same at the end of the function.

File: src/vehicle.cpp (hunk 2 above)
Comment (mugling) [inline src/vehicle.cpp]: No good reason and considering I wrote both implementations no good excuse either.

File: src/vehicle.cpp (hunk 3 above)
Comment (mugling) [inline src/vehicle.cpp]: The line count isn't especially relevant. I'd argue it's actually clearer? Not especially attached to it though.

File: src/vehicle.cpp [hunk 4]
```diff
@@ -4195,34 +4139,30 @@ void vehicle::alarm() {
...
-            }
-            part.amount -= leak_amount;
+
+        auto fuel = p.ammo_current();
+        if( fuel != fuel_type_gasoline && fuel != fuel_type_diesel &&
+            fuel != fuel_type_battery && fuel != fuel_type_water ) {
+            continue; // not a liquid fuel or battery
+        }
+
+        int qty = std::min( ( 0.5 - dmg ) * ( 0.5 - dmg) * p.ammo_remaining() / 10, 1.0 );
```
Comment (Coolthulhu) [inline src/vehicle.cpp]: This inverts the original logic, which was to low cap leak amount - now it's an upper cap.

**Comment by Coolthulhu** [comment]
When manually firing a vehicle-mounted laser rifle, it debugmsgd about "trying to set invalid ammo of none for fake UPS".

**Comment by mugling** [comment]
> When manually firing a vehicle-mounted laser rifle, it debugmsgd about "trying to set invalid ammo of none for fake UPS".

Existing bug

**Comment by Coolthulhu** [comment]
The turrets also seem not to be able to select own targets. Once I mark one manually, they'll keep shooting at it until they can no longer do so, then they'll "lose aggro".

**Pull request merged**